clients/drcachesim/tests/instr_decode_cache_test.cpp

```diff
@@ -0,0 +1,167 @@
+    instr_t *ret = XINST_CREATE_return(GLOBAL_DCONTEXT);
+    instrlist_t *ilist = instrlist_create(GLOBAL_DCONTEXT);
+    instrlist_append(ilist, nop);
+    instrlist_append(ilist, ret);
+    std::vector<memref_with_IR_t> memref_setup = {
+        { gen_instr(TID_A), nop },
+        { gen_instr(TID_A), ret },
+        { gen_instr(TID_A), nop },
+    };
+    auto memrefs = add_encodings_to_memrefs(ilist, memref_setup, BASE_ADDR);
+    // Set up the second nop memref to reuse the same encoding as the first nop.
+    memrefs[2].instr.encoding_is_new = false;
+
+    instr_decode_cache_t<test_decode_info_t> decode_cache(
+        GLOBAL_DCONTEXT,
+        /*persist_decoded_instrs=*/false);
+    if (decode_cache.get_decode_info(reinterpret_cast<app_pc>(memrefs[0].instr.addr)) !=
+        nullptr) {
+        return "Unexpected decode info for never-seen pc";
+    }
+    decode_cache.add_decode_info(memrefs[0].instr);
+    test_decode_info_t *decode_info_nop =
+        decode_cache.get_decode_info(reinterpret_cast<app_pc>(memrefs[0].instr.addr));
+    if (decode_info_nop == nullptr || !decode_info_nop->is_nop) {
+        return "Unexpected decode info for nop instr";
+    }
+    decode_cache.add_decode_info(memrefs[1].instr);
+    test_decode_info_t *decode_info_ret =
+        decode_cache.get_decode_info(reinterpret_cast<app_pc>(memrefs[1].instr.addr));
+    if (decode_info_ret == nullptr || !decode_info_ret->is_ret) {
+        return "Unexpected decode info for ret instr";
+    }
+    decode_cache.add_decode_info(memrefs[2].instr);
+    test_decode_info_t *decode_info_nop_2 =
+        decode_cache.get_decode_info(reinterpret_cast<app_pc>(memrefs[2].instr.addr));
+    if (decode_info_nop_2 != decode_info_nop) {
+        return "Did not see same decode info instance for second instance of nop";
+    }
+
+    instrlist_clear_and_destroy(GLOBAL_DCONTEXT, ilist);
+    std::cerr << "check_decode_caching_without_instr passed\n";
+    return "";
+}
+
+std::string
+check_instr_decode_caching()
+{
+    static constexpr addr_t BASE_ADDR = 0x123450;
+    static constexpr addr_t TID_A = 1;
+    instr_t *nop = XINST_CREATE_nop(GLOBAL_DCONTEXT);
+    instr_t *ret = XINST_CREATE_return(GLOBAL_DCONTEXT);
+    instrlist_t *ilist = instrlist_create(GLOBAL_DCONTEXT);
+    instrlist_append(ilist, nop);
+    instrlist_append(ilist, ret);
+    std::vector<memref_with_IR_t> memref_setup = {
+        { gen_instr(TID_A), nop },
+        { gen_instr(TID_A), ret },
+        { gen_instr(TID_A), nop },
+    };
+    auto memrefs = add_encodings_to_memrefs(ilist, memref_setup, BASE_ADDR);
+    // Set up the second nop memref to reuse the same encoding as the first nop.
+    memrefs[2].instr.encoding_is_new = false;
+
+    instr_decode_cache_t<instr_decode_info_t> decode_cache(
+        GLOBAL_DCONTEXT,
+        /*persist_decoded_instr=*/true);
+    for (const memref_t &memref : memrefs) {
+        decode_cache.add_decode_info(memref.instr);
```

Comment:
Wait, why is this one different: there's no real reason, right?  Could this test in fact use the same incremental adding approach of the test above?  I.e., could we delete this entire function and make a parameter to check_decode_caching_without_instr() (renamed minus "_without_instr") for `persist_decoded_instr` and have all the checks be identical?  Just call it twice, once false and once true?   That would be clearer code and more sharing.  Otherwise the reader thinks there's some reason this variant has to add all the decode info up front, or some reason it doesn't want to do the extra checks of the other one.

Reply:
> Could this test in fact use the same incremental adding approach of the test above?

That doesn't lend any additional test coverage. The incremental-adding of the test above already verifies the correct operation of the cache, so we don't need to repeat it here.

This test is intended to verify the instr_t stored by instr_decode_cache_t. So both unit tests have different intent.

It makes sense to me to keep the checks separate, though they can share the setup code (I previously decided on not sharing the setup code in favor of keeping the tests separate as their intents are different; I'll add a comment to note that).

Reply:
But if the checks in this 2nd test are a strict subset of the first, why write separate checks at all?  Not convinced it's not better to just share the 1st test's checks.

Reply:
> But if the checks in this 2nd test are a strict subset of the first, why write separate checks at all?

They're not.

The 1st test is for instr_decode_cache_t, whereas the 2nd one is mostly for instr_decode_info_t.

Reply:
Shared the setup code.

Reply:
OK, they are different, the type names were all blurring together I think (combos of "decode" and "info" and "test" in different orders).